File: proof/browser/body-journey-track.mjs
import { createHash } from "node:crypto";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";

const STEPS = [
  ["body.absent", "body-absent", "runtime-receipt"], ["bootstrap.started", "bootstrap-started", "runtime-receipt"],
  ["body.born", "body-born", "body-biography"], ["body.awake", "body-awake", "body-biography"],
  ["form.used", "standing-form-used", "runtime-receipt"], ["body.inspected", "body-inspected", "semantic-presentation"],
  ["workload.revised", "workload-revised", "body-biography"], ["host.added", "host-added", "runtime-receipt"],
  ["fault.observed", "fault-observed", "stream-disposition"], ["body.repaired", "body-repaired", "runtime-receipt"],
  ["body.long-running", "body-long-running", "runtime-receipt"], ["body.lulled", "body-lulled", "lifecycle-action"],
  ["body.fulfilled", "body-fulfilled", "fulfilled-transition"],
];
const digest = bytes => `sha256:${createHash("sha256").update(bytes).digest("hex")}`;
const required = (value, label) => {
  if (typeof value !== "string" || value.length === 0) throw new Error(`browser track lacks ${label}`);
  return value;
};

export async function writeBrowserBodyJourneyTrack(source, screenshotPath, output, captures = {}, videoPath) {
  const commit = required(source.git_commit, "exact commit");
  if (!/^[0-9a-f]{40}$/.test(commit)) throw new Error("browser track refuses a non-exact commit");
  const current = source.current;
  const evidence = source.evidence.evidence;
  const events = evidence.body.events;
  const joined = source.checkpoints.joined.evidence.membership.parts.map(part => part.current).filter(Boolean);
  const peer = joined.find(host => host.host_id !== current.host_id);
  const refusedWake = source.checkpoints.refused.evidence.evidence.wakes.at(-1);
  const repairedWake = source.checkpoints.repaired.evidence.evidence.wakes.at(-1);
  const plan = repairedWake.plans[0];
  const born = required(events[0]?.Born?.sign_id, "birth sign");
  const fulfilled = required(events.at(-1)?.Fulfilled?.sign_id, "fulfillment sign");
  const wake = required(evidence.wakes[0]?.sign_ids?.[0], "Wake sign");
  const workload = required(events.find(event => event.FormAdmitted)?.FormAdmitted?.sign_id, "workload sign");
  const fault = required(refusedWake.sign_ids.at(-1), "fault sign");
  const repaired = required(repairedWake.sign_ids.at(-1), "repair sign");
  const lull = required(events.findLast(event => event.LullRetained)?.LullRetained?.sign_id, "Lull sign");
  const manifestation = digest(await readFile(screenshotPath));
  const ids = { body: required(current.body_id, "Body identity"), host: required(current.host_id, "Host identity"),
    boot: required(current.boot_id, "Boot identity"), peerHost: required(peer?.host_id, "peer Host identity"),
    peerBoot: required(peer?.boot_id, "peer Boot identity"), plan: required(plan.plan_id, "Plan identity"),
    play: required(plan.active_play_id, "Play identity"),
    presentation: required(source.state.terminal.presentation_id, "Presentation identity"), manifestation };
  const facts = [
    { initial_body: null, host_id: ids.host, boot_id: ids.boot }, { host_id: ids.host, boot_id: ids.boot },
    { event: events[0] }, { wake: evidence.wakes[0] }, { plan_id: ids.plan, active_play_id: ids.play },
    { presentation_id: ids.presentation, manifestation_id: manifestation },
    { workload_revision: current.workload_revision, forms: evidence.body.workset.forms },
    { peer, membership_revision: source.checkpoints.joined.evidence.membership.revision },
    { refusal: source.checkpoints.refused.playback.refusal, wake: refusedWake }, { wake: repairedWake },
    { active_play_id: ids.play }, { lull_sign_id: lull }, { event: events.at(-1) },
  ];
  const signs = [null, null, born, wake, repaired, null, workload,
    required(source.checkpoints.joined.evidence.membership.events.at(-1)?.sign_id, "join sign"),
    fault, repaired, repaired, lull, fulfilled];
  await mkdir(join(output, "artifacts"), { recursive: true });
  const steps = [];
  for (let index = 0; index < STEPS.length; index += 1) {
    const [stepId, assertion, rung] = STEPS[index];
    const relative = `artifacts/${String(index + 1).padStart(2, "0")}-${stepId}.json`;
    const bytes = Buffer.from(`${JSON.stringify({ schema: "conduit.evidence/semantic-step-receipt@2",
      git_commit: commit, track: "browser-graphical", step_id: stepId, assertion,
      source_facts: facts[index] }, null, 2)}\n`);
    await writeFile(join(output, relative), bytes, { flag: "wx" });
    const hostAdded = stepId === "host.added";
    steps.push({ step_id: stepId, assertion, disposition: "established", provenance: {
      body_id: index >= 2 ? ids.body : null, host_id: index < 3 ? ids.host : hostAdded ? ids.peerHost : null,
      boot_id: index < 3 ? ids.boot : hostAdded ? ids.peerBoot : null,
      plan_id: ["form.used", "workload.revised", "host.added", "body.repaired"].includes(stepId) ? ids.plan : null,
      play_id: ["form.used", "body.repaired", "body.long-running"].includes(stepId) ? ids.play : null,
      presentation_id: stepId === "body.inspected" ? ids.presentation : null,
      manifestation_id: stepId === "body.inspected" ? ids.manifestation : null, line_id: null, sign_id: signs[index],
    }, evidence: [{ artifact_id: `browser-graphical/${stepId}`, evidence_class: "semantic-receipt",
      assertion_rung: rung, documentary_description: `browser-wasm-body producer receipt for ${stepId}.`,
      path: relative, sha256: digest(bytes) }] });
  }
  for (const [stepId, capture] of Object.entries(captures)) {
    const step = steps.find(candidate => candidate.step_id === stepId);
    if (!step) throw new Error(`unknown captured journey step: ${stepId}`);
    const bytes = await readFile(capture.path);
    if (!bytes.subarray(0, 8).equals(Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]))) {
      throw new Error(`journey screenshot is not PNG: ${stepId}`);
    }
    const relative = `artifacts/${stepId}.png`;
    await writeFile(join(output, relative), bytes, { flag: "wx" });
    step.evidence.push({ artifact_id: `browser-graphical/${stepId}/screen`,
      evidence_class: "screenshot", assertion_rung: "deterministic-observation",
      documentary_description: capture.caption, path: relative, sha256: digest(bytes) });
  }
  if (videoPath) {
    const bytes = await readFile(videoPath);
    const relative = "artifacts/browser-body-session.webm";
    await writeFile(join(output, relative), bytes, { flag: "wx" });
    steps.at(-1).evidence.push({ artifact_id: "browser-graphical/session-video",
      evidence_class: "video", assertion_rung: "deterministic-observation",
      documentary_description: "Watch the complete uncut browser session, including birth, host admission, refusal, recovery and fulfillment.",
      path: relative, sha256: digest(bytes) });
  }
  await writeFile(join(output, "track.json"), `${JSON.stringify({
    schema: "conduit.evidence/body-journey-track@2", journey_id: "orifina/tutorial@1", git_commit: commit,
    track_id: "browser-graphical", embodiment: "browser-wasm-body", body_id: ids.body,
    presenter_id: "presenter/browser-dom@1", hosts: [
      { host_id: ids.host, boot_id: ids.boot }, { host_id: ids.peerHost, boot_id: ids.peerBoot },
    ], line_ids: [], distributed_plan_ids: [], steps,
  }, null, 2)}\n`, { flag: "wx" });
}

File: tools/xtask/src/commands/body_journey_track.rs
//! Producer-side retention for one lived Body tutorial track.

use serde::Serialize;
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::{fs, path::Path};

const STEPS: [(&str, &str, &str); 13] = [
    ("body.absent", "body-absent", "runtime-receipt"),
    ("bootstrap.started", "bootstrap-started", "runtime-receipt"),
    ("body.born", "body-born", "body-biography"),
    ("body.awake", "body-awake", "body-biography"),
    ("form.used", "standing-form-used", "runtime-receipt"),
    ("body.inspected", "body-inspected", "semantic-presentation"),
    ("workload.revised", "workload-revised", "body-biography"),
    ("host.added", "host-added", "runtime-receipt"),
    ("fault.observed", "fault-observed", "stream-disposition"),
    ("body.repaired", "body-repaired", "runtime-receipt"),
    ("body.long-running", "body-long-running", "runtime-receipt"),
    ("body.lulled", "body-lulled", "lifecycle-action"),
    ("body.fulfilled", "body-fulfilled", "fulfilled-transition"),
];

pub(crate) struct TrackIdentities {
    pub body: String,
    pub host: String,
    pub boot: String,
    pub peer_host: String,
    pub peer_boot: String,
    pub plan: String,
    pub distributed_plan: Option<String>,
    pub play: String,
    pub presentation: String,
    pub manifestation: String,
    pub line: Option<String>,
    pub signs: [Option<String>; 13],
}

pub(crate) struct TrackSource {
    pub commit: String,
    pub track_id: &'static str,
    pub embodiment: &'static str,
    pub presenter_id: &'static str,
    pub identities: TrackIdentities,
    pub facts: [Value; 13],
}

#[derive(Serialize)]
struct Artifact<'a> {
    schema: &'static str,
    git_commit: &'a str,
    track: &'a str,
    step_id: &'a str,
    assertion: &'a str,
    source_facts: &'a Value,
}

pub(crate) fn write(source: TrackSource, output: &Path) -> Result<(), String> {
    if source.commit.len() != 40 || !source.commit.bytes().all(|byte| byte.is_ascii_hexdigit()) {
        return Err("Body track requires an exact commit".into());
    }
    let artifacts = output.join("artifacts");
    fs::create_dir_all(&artifacts)
        .map_err(|error| format!("create Body track artifacts: {error}"))?;
    let mut steps = Vec::with_capacity(STEPS.len());
    for (index, ((step_id, assertion, rung), facts)) in
        STEPS.iter().zip(source.facts.iter()).enumerate()
    {
        let file = format!("{:02}-{step_id}.json", index + 1);
        let relative = format!("artifacts/{file}");
        let bytes = serde_json::to_vec_pretty(&Artifact {
            schema: "conduit.evidence/semantic-step-receipt@2",
            git_commit: &source.commit,
            track: source.track_id,
            step_id,
            assertion,
            source_facts: facts,
        })
        .map_err(|error| format!("encode {step_id} receipt: {error}"))?;
        write_new(&artifacts.join(file), &bytes)?;
        let host_added = *step_id == "host.added";
        let plan = matches!(
            *step_id,
            "form.used" | "workload.revised" | "host.added" | "body.repaired"
        )
        .then(|| {
            if host_added {
                source
                    .identities
                    .distributed_plan
                    .as_ref()
                    .unwrap_or(&source.identities.plan)
            } else {
                &source.identities.plan
            }
        });
        let play = matches!(
            *step_id,
            "form.used" | "body.repaired" | "body.long-running"
        )
        .then_some(&source.identities.play);
        let sign = source.identities.signs[index].as_ref();
        steps.push(serde_json::json!({
            "step_id": step_id,
            "assertion": assertion,
            "disposition": "established",
            "provenance": {
                "body_id": (index >= 2).then_some(&source.identities.body),
                "host_id": if index < 3 { Some(&source.identities.host) } else if host_added { Some(&source.identities.peer_host) } else { None },
                "boot_id": if index < 3 { Some(&source.identities.boot) } else if host_added { Some(&source.identities.peer_boot) } else { None },
                "plan_id": plan,
                "play_id": play,
                "presentation_id": (*step_id == "body.inspected").then_some(&source.identities.presentation),
                "manifestation_id": (*step_id == "body.inspected").then_some(&source.identities.manifestation),
                "line_id": (host_added).then_some(source.identities.line.as_ref()).flatten(),
                "sign_id": sign,
            },
            "evidence": [{
                "artifact_id": format!("{}/{step_id}", source.track_id),
                "evidence_class": "semantic-receipt",
                "assertion_rung": rung,
                "documentary_description": format!("{} producer receipt for {step_id}.", source.embodiment),
                "path": relative,
                "sha256": format!("sha256:{:x}", Sha256::digest(&bytes)),
            }],
        }));
    }
    let mut hosts = vec![serde_json::json!({
        "host_id": source.identities.host,
        "boot_id": source.identities.boot,
    })];
    hosts.push(serde_json::json!({
        "host_id": source.identities.peer_host,
        "boot_id": source.identities.peer_boot,
    }));
    let document = serde_json::json!({
        "schema": "conduit.evidence/body-journey-track@2",
        "journey_id": "orifina/tutorial@1",
        "git_commit": source.commit,
        "track_id": source.track_id,
        "embodiment": source.embodiment,
        "body_id": source.identities.body,
        "presenter_id": source.presenter_id,
        "hosts": hosts,
        "line_ids": source.identities.line.into_iter().collect::<Vec<_>>(),
        "distributed_plan_ids": source.identities.distributed_plan.into_iter().collect::<Vec<_>>(),
        "steps": steps,
    });
    let bytes = serde_json::to_vec_pretty(&document)
        .map_err(|error| format!("encode Body track: {error}"))?;
    write_new(&output.join("track.json"), &bytes)
}

fn write_new(path: &Path, bytes: &[u8]) -> Result<(), String> {
    use std::io::Write;
    fs::OpenOptions::new()
        .write(true)
        .create_new(true)
        .open(path)
        .and_then(|mut file| file.write_all(bytes))
        .map_err(|error| format!("create {}: {error}", path.display()))
}
